geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -2176,9 +2175,8 @@ public void close(String reason, Throwable systemFailureCause, boolean keepAlive
           prepareDiskStoresForClose();
 
           List<InternalRegion> rootRegionValues;
-          synchronized (rootRegions) {
-            rootRegionValues = new ArrayList<>(rootRegions.values());
-          }
+
+          rootRegionValues = new ArrayList<>(rootRegions.values());
```

Comment:
I see no reason for continuing to have this method make a copy of "rootRegions.values()". All it does with "rootRegionValues" is an iteration. So now that you have changed to a ConcurrentMap, I think you can safely get rid of "rootRegionValues" and down on line 2194 do this: for (InternalRegion lr: rootRegions.values())

Reply:
If we do `for (InternalRegion lr : rootRegions.vales())`, other threads may change `rootRegions` while the `for` loop is still in progress. Since each `for` iteration evaluates `rootRegions.values()`, the resulting behavior could be different than the original behavior, which iterates over a snapshot of `rootRegions`.

Reply:
I you think about what you are doing now when you make the copy, it is actually iterating over the values to copy them into your ArrayList; without any type of lock. Basically ConcurrentMap gives you a Set that you can safely iterate without locking. It may change but you would only notice this if you did something like iterate it twice and compared the first iteration to the second. Since all this code does is a simple iteration you don't need to make a copy to be safe. It is not going to break the code; but you are wasting some memory and cpu and possibly making future coders think you need to make a copy to be safe :-)

Reply:
By the way, each iteration does not evaluate rootRegion.values(). Values is called once and the result is iterated. Each iteration gets the next value from the Set that values() returned.